flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java

```diff
@@ -1060,40 +1066,24 @@ protected CleanupJobState jobReachedTerminalState(ExecutionGraphInfo executionGr
                     terminalJobStatus);
         }
 
-        archiveExecutionGraph(executionGraphInfo);
+        storeExecutionGraphInfo(executionGraphInfo);
 
         if (terminalJobStatus.isGloballyTerminalState()) {
-            final JobID jobId = executionGraphInfo.getJobId();
-            try {
-                if (jobResultStore.hasCleanJobResultEntry(jobId)) {
-                    log.warn(
-                            "Job {} is already marked as clean but clean up was triggered again.",
-                            jobId);
-                } else if (!jobResultStore.hasDirtyJobResultEntry(jobId)) {
-                    jobResultStore.createDirtyResult(
-                            new JobResultEntry(
-                                    JobResult.createFrom(
-                                            executionGraphInfo.getArchivedExecutionGraph())));
-                    log.info(
-                            "Job {} has been registered for cleanup in the JobResultStore after reaching a terminal state.",
-                            jobId);
-                }
-            } catch (IOException e) {
-                fatalErrorHandler.onFatalError(
-                        new FlinkException(
-                                String.format(
-                                        "The job %s couldn't be marked as pre-cleanup finished in JobResultStore.",
-                                        jobId),
-                                e));
-            }
-        }
 
-        return terminalJobStatus.isGloballyTerminalState()
-                ? CleanupJobState.GLOBAL
-                : CleanupJobState.LOCAL;
+            // do not create an archive for suspended jobs, as this would eventually lead to
+            // multiple archive attempts which we currently do not support
+            CompletableFuture<Acknowledge> archiveFuture =
+                    archiveExecutionGraph(executionGraphInfo);
+
+            registerCleanupInJobResultStore(executionGraphInfo);
```

Comment:
No, `thenApply` only works in the successful case. We want to register the job as completed even if the archiving failed. Therefore, using `whenComplete` is the better option.

> However, I put it here because I think if the archiving takes a long time and the cluster is killed by the user or the external resource providers, the registration is done already and the cleanup will be executed once the job is resumed.

The reasoning behind chaining them together was, that the HistoryServer is opt-in functionality, i.e. if the user configured it, he/she would expect it to work even if it takes time. The default works with a no-op history server. WDYT?

Reply:
> No, thenApply only works in the successful case. We want to register the job as completed even if the archiving failed. Therefore, using whenComplete is the better option.

The exceptions in `archiveExecutionGraphToHistoryServer` has been handled by the code inside it:
```java
return executionGraphFuture.handle(
                (Acknowledge ignored, Throwable throwable) -> {
                    if (throwable != null) {
                        log.info(
                                "Could not archive completed job {}({}) to the history server.",
                                executionGraphInfo.getArchivedExecutionGraph().getJobName(),
                                executionGraphInfo.getArchivedExecutionGraph().getJobID(),
                                throwable);
                    }
                    return Acknowledge.get();
                });
```
In this way, `submitFailedJob` doesn't need to handle the exception again. 

That's why I think putting the registration into `thenApplyAsync` works. WDYT?

> The reasoning behind chaining them together was, that the HistoryServer is opt-in functionality, i.e. if the user configured it, he/she would expect it to work even if it takes time. The default works with a no-op history server. WDYT?

Yes, you are right. Sorry for making you rephrase it over and over again for my information.

Reply:
Fair enough, but do you see any problems with using `whenComplete`? My concern is just, that using `thenApply` explicitly states that the registration in the JobResultStore should only happen in the case of success. Assuming knowledge about how `archiveExecutionGraphToHistoryServer` is valid because it's a private method of `Dispatcher`. But it makes reading the code harder because the reader really has to know that `archiveExecutionGraphToHistoryServer` always returns a successfully completed future.

Reply:
Yes, you are right. We cannot have the assumption that `archiveExecutionGraphToHistoryServer` always returns a successfully completed future. How about `handleAsync`? In this way, the code explicitly indicates that `registerGloballyTerminatedJobInJobResultStore` happens before `return CleanupJobState.GLOBAL`.

```java
CompletableFuture<Acknowledge> archiveFuture =
    archiveExecutionGraphToHistoryServer(executionGraphInfo);

return archiveFuture.handleAsync(
    (ignored, throwable) -> {
        registerGloballyTerminatedJobInJobResultStore(executionGraphInfo);
        return CleanupJobState.GLOBAL;
     },
getMainThreadExecutor());
```